File: python/src/raidxai/resources/images.py
"""Image forensics — detect AI-generated, deepfaked, and digitally edited images. Synchronous."""

from __future__ import annotations

from collections.abc import Callable
from typing import TYPE_CHECKING

from .. import _specs
from .._poll import poll_until, poll_until_async
from ..files import FileInput
from ..models import (
    ImageBatch,
    ImageBatchCancelResult,
    ImageBatchPage,
    ImageBatchResultPage,
    ImageBatchSubmitResponse,
    ImageForensicsCreditInfo,
    ImageForensicsResponse,
)

if TYPE_CHECKING:
    from ..aio import AsyncRaidClient
    from ..client import RaidClient


def _to_list(files: FileInput | list[FileInput]) -> list[FileInput]:
    return files if isinstance(files, list) else [files]


class Images:
    def __init__(self, client: RaidClient) -> None:
        self._client = client

    def process(
        self,
        files: FileInput | list[FileInput],
        *,
        external_id: str | None = None,
        source_url: str | None = None,
    ) -> ImageForensicsResponse:
        """Analyze one or more image files (up to 10). Requires the ``image`` scope."""
        spec = _specs.image_process(_to_list(files), external_id, source_url)
        return self._client.request(spec, ImageForensicsResponse)

    def process_from_url(
        self, url: str, *, external_id: str | None = None
    ) -> ImageForensicsResponse:
        """Analyze an image fetched from a public URL."""
        spec = _specs.image_process_from_url(url, external_id)
        return self._client.request(spec, ImageForensicsResponse)

    def submit_batch(
        self,
        files: list[FileInput],
        *,
        external_id: str | None = None,
    ) -> ImageBatchSubmitResponse:
        """Queue images for analysis in a later scheduled window, at a reduced credit cost.

        Returns immediately with a batch id -- **no verdicts**. Credits are deducted on
        acceptance and refunded for any image that ultimately fails.

        Batch mode is images-only; no other modality has a batch tier.

        Two differences from :meth:`process` are structural, not temporary:

        * **Turnaround is up to a full processing window**, so never put this on a path where a
          person is waiting.
        * **Results are basic tier** -- ``deep_analysis``, ``generators`` and ``heatmap_url`` are
          always ``None``, whatever the plan. Those are realtime-only.
        """
        spec = _specs.image_submit_batch(files, external_id=external_id)
        return self._client.request(spec, ImageBatchSubmitResponse)

    def get_batch(self, batch_id: str) -> ImageBatch:
        """Current status and counters for one batch."""
        return self._client.request(_specs.image_get_batch(batch_id), ImageBatch)

    def list_batches(self, *, skip: int | None = None, take: int | None = None) -> ImageBatchPage:
        """Your batches, newest first."""
        return self._client.request(_specs.image_list_batches(skip=skip, take=take), ImageBatchPage)

    def batch_results(
        self, batch_id: str, *, skip: int | None = None, take: int | None = None
    ) -> ImageBatchResultPage:
        """Per-image results.

        Individual images become available as they finish -- you do not have to wait for the whole
        batch. Match results to your inputs with ``ordinal``, the image's position in the original
        submission; file names are echoed back but are not guaranteed unique.
        """
        spec = _specs.image_batch_results(batch_id, skip=skip, take=take)
        return self._client.request(spec, ImageBatchResultPage)

    def cancel_batch(self, batch_id: str) -> ImageBatchCancelResult:
        """Cancel the images still waiting and refund their credits.

        Images already being analyzed are left to finish -- they have consumed the work either
        way -- so the refund covers only what was actually cancelled.
        """
        return self._client.request(_specs.image_cancel_batch(batch_id), ImageBatchCancelResult)

    def submit_batch_and_wait(
        self,
        files: list[FileInput],
        *,
        external_id: str | None = None,
        interval_seconds: float = 3.0,
        timeout_seconds: float = 300.0,
        on_poll: Callable[[ImageBatch], None] | None = None,
    ) -> ImageBatch:
        """Queue a batch and poll until it is terminal, then return it.

        **Rarely what you want.** A batch can legitimately take hours, so this holds the process
        open for the whole window and the default timeout will usually expire first. Prefer
        :meth:`submit_batch`, persist the id, and collect results later -- that is the point of
        the tier. This exists for tests and short-window setups.

        Polling stops on the server-computed ``is_terminal`` rather than a status allow-list, so
        it keeps working if a status is ever added.
        """
        submitted = self.submit_batch(files, external_id=external_id)
        batch_id = submitted.batch_id
        if batch_id is None:
            raise ValueError("submit_batch did not return a batch_id")

        return poll_until(
            lambda: self.get_batch(str(batch_id)),
            lambda b: b.is_terminal is True,
            lambda b: b.status.value if b.status else "unknown",
            interval_seconds=interval_seconds,
            timeout_seconds=timeout_seconds,
            on_poll=on_poll,
        )

    def credit_info(self) -> ImageForensicsCreditInfo:
        """Read the credit balance and the live image limits: cost per image, max file size and
        allowed formats, plus whether the batch tier is available and what it costs.

        Read these rather than hard-coding the defaults — they are account settings and can
        change without an SDK release. Any valid API token may call this; no scope is required.
        """
        return self._client.request(_specs.image_credit_info(), ImageForensicsCreditInfo)


class AsyncImages:
    def __init__(self, client: AsyncRaidClient) -> None:
        self._client = client

    async def process(
        self,
        files: FileInput | list[FileInput],
        *,
        external_id: str | None = None,
        source_url: str | None = None,
    ) -> ImageForensicsResponse:
        spec = _specs.image_process(_to_list(files), external_id, source_url)
        return await self._client.request(spec, ImageForensicsResponse)

    async def process_from_url(
        self, url: str, *, external_id: str | None = None
    ) -> ImageForensicsResponse:
        spec = _specs.image_process_from_url(url, external_id)
        return await self._client.request(spec, ImageForensicsResponse)


    async def submit_batch(
        self,
        files: list[FileInput],
        *,
        external_id: str | None = None,
    ) -> ImageBatchSubmitResponse:
        """Queue images for analysis in a later scheduled window, at a reduced credit cost.

        Returns immediately with a batch id -- **no verdicts**. Credits are deducted on
        acceptance and refunded for any image that ultimately fails.

        Batch mode is images-only; no other modality has a batch tier.

        Two differences from :meth:`process` are structural, not temporary:

        * **Turnaround is up to a full processing window**, so never put this on a path where a
          person is waiting.
        * **Results are basic tier** -- ``deep_analysis``, ``generators`` and ``heatmap_url`` are
          always ``None``, whatever the plan. Those are realtime-only.
        """
        spec = _specs.image_submit_batch(files, external_id=external_id)
        return await self._client.request(spec, ImageBatchSubmitResponse)

    async def get_batch(self, batch_id: str) -> ImageBatch:
        """Current status and counters for one batch."""
        return await self._client.request(_specs.image_get_batch(batch_id), ImageBatch)

    async def list_batches(
        self, *, skip: int | None = None, take: int | None = None
    ) -> ImageBatchPage:
        """Your batches, newest first."""
        spec = _specs.image_list_batches(skip=skip, take=take)
        return await self._client.request(spec, ImageBatchPage)

    async def batch_results(
        self, batch_id: str, *, skip: int | None = None, take: int | None = None
    ) -> ImageBatchResultPage:
        """Per-image results.

        Individual images become available as they finish -- you do not have to wait for the whole
        batch. Match results to your inputs with ``ordinal``, the image's position in the original
        submission; file names are echoed back but are not guaranteed unique.
        """
        spec = _specs.image_batch_results(batch_id, skip=skip, take=take)
        return await self._client.request(spec, ImageBatchResultPage)

    async def cancel_batch(self, batch_id: str) -> ImageBatchCancelResult:
        """Cancel the images still waiting and refund their credits.

        Images already being analyzed are left to finish -- they have consumed the work either
        way -- so the refund covers only what was actually cancelled.
        """
        spec = _specs.image_cancel_batch(batch_id)
        return await self._client.request(spec, ImageBatchCancelResult)

    async def submit_batch_and_wait(
        self,
        files: list[FileInput],
        *,
        external_id: str | None = None,
        interval_seconds: float = 3.0,
        timeout_seconds: float = 300.0,
        on_poll: Callable[[ImageBatch], None] | None = None,
    ) -> ImageBatch:
        """Queue a batch and poll until it is terminal, then return it.

        **Rarely what you want.** A batch can legitimately take hours, so this holds the process
        open for the whole window and the default timeout will usually expire first. Prefer
        :meth:`submit_batch`, persist the id, and collect results later -- that is the point of
        the tier. This exists for tests and short-window setups.

        Polling stops on the server-computed ``is_terminal`` rather than a status allow-list, so
        it keeps working if a status is ever added.
        """
        submitted = await self.submit_batch(files, external_id=external_id)
        batch_id = submitted.batch_id
        if batch_id is None:
            raise ValueError("submit_batch did not return a batch_id")

        return await poll_until_async(
            lambda: self.get_batch(str(batch_id)),
            lambda b: b.is_terminal is True,
            lambda b: b.status.value if b.status else "unknown",
            interval_seconds=interval_seconds,
            timeout_seconds=timeout_seconds,
            on_poll=on_poll,
        )

    async def credit_info(self) -> ImageForensicsCreditInfo:
        return await self._client.request(_specs.image_credit_info(), ImageForensicsCreditInfo)

File: typescript/src/resources/images.ts
import type { RaidClient } from "../client.js";
import type {
  ImageBatch,
  ImageBatchCancelResult,
  ImageBatchPage,
  ImageBatchResultPage,
  ImageBatchSubmitResponse,
  ImageForensicsCreditInfo,
  ImageForensicsResponse,
} from "../types.js";
import { pollUntil, type PollOptions } from "../poll.js";
import { appendFile, type FileInput } from "../upload.js";

export interface ImageProcessOptions {
  /** Caller-provided ID to correlate this analysis with your own systems. */
  externalId?: string;
  /** Source URL of the image, recorded for your own auditing. */
  sourceUrl?: string;
}

export interface ImageProcessFromUrlOptions {
  externalId?: string;
}

/** Image forensics — detect AI-generated, deepfaked, and digitally edited images. Synchronous. */

/** Options for {@link Images.submitBatch}. */
export interface ImageBatchSubmitOptions {
  /** Caller-provided ID echoed back on the batch, to reconcile against your own systems. */
  externalId?: string;
}

/** Paging options for {@link Images.listBatches} and {@link Images.batchResults}. */
export interface ImageBatchListOptions {
  skip?: number;
  take?: number;
}

export class Images {
  constructor(private readonly client: RaidClient) {}

  /**
   * Analyze one or more image files (up to 10 per request). Returns a verdict,
   * confidence, and — on detailed plans — generator attribution per image.
   * Requires an API key with the `image` scope.
   */
  process(files: FileInput | FileInput[], options: ImageProcessOptions = {}): Promise<ImageForensicsResponse> {
    const list = Array.isArray(files) ? files : [files];
    if (list.length === 0) {
      throw new Error("images.process: provide at least one image file.");
    }
    const form = new FormData();
    for (const file of list) {
      appendFile(form, "imageFiles", file);
    }
    if (options.externalId !== undefined) form.append("input.ExternalId", options.externalId);
    if (options.sourceUrl !== undefined) form.append("sourceUrl", options.sourceUrl);

    return this.client.request<ImageForensicsResponse>({
      method: "POST",
      path: "/api/app/image-forensics/process",
      form,
    });
  }

  /** Analyze an image fetched from a public URL (a direct link or a page the resolver extracts from). */
  processFromUrl(url: string, options: ImageProcessFromUrlOptions = {}): Promise<ImageForensicsResponse> {
    return this.client.request<ImageForensicsResponse>({
      method: "POST",
      path: "/api/app/image-forensics/process-from-url",
      json: { url, externalId: options.externalId },
    });
  }

  /**
   * Queue images for analysis in a later scheduled window, at a reduced credit cost.
   *
   * Returns immediately with a batch ID — **no verdicts**. Credits are deducted on acceptance,
   * and refunded for any image that ultimately fails.
   *
   * Batch mode is images-only; no other modality has a batch tier.
   *
   * Two differences from {@link Images.process} are structural, not temporary:
   * - **Turnaround is up to a full processing window**, so never put this on a path where a
   *   person is waiting.
   * - **Results are basic tier** — `deepAnalysis`, `generators` and `heatmapUrl` are always null,
   *   whatever the plan. Those are realtime-only.
   */
  async submitBatch(
    files: FileInput[],
    options: ImageBatchSubmitOptions = {},
  ): Promise<ImageBatchSubmitResponse> {
    if (files.length === 0) {
      throw new Error("submitBatch requires at least one image");
    }

    const form = new FormData();
    for (const file of files) {
      appendFile(form, "imageFiles", file);
    }
    if (options.externalId) form.append("input.ExternalId", options.externalId);

    return this.client.request<ImageBatchSubmitResponse>({
      method: "POST",
      path: "/api/app/image-forensics/batches",
      form,
    });
  }

  /** Current status and counters for one batch. */
  async getBatch(batchId: string): Promise<ImageBatch> {
    return this.client.request<ImageBatch>({
      method: "GET",
      path: `/api/app/image-forensics/batches/${encodeURIComponent(batchId)}`,
    });
  }

  /** Your batches, newest first. */
  async listBatches(options: ImageBatchListOptions = {}): Promise<ImageBatchPage> {
    return this.client.request<ImageBatchPage>({
      method: "GET",
      path: "/api/app/image-forensics/batches",
      query: { skip: options.skip, take: options.take },
    });
  }

  /**
   * Per-image results.
   *
   * Individual images become available as they finish — you do not have to wait for the whole
   * batch. Match results to your inputs with `ordinal`, which is the image's position in the
   * original submission; file names are echoed back but are not guaranteed unique.
   */
  async batchResults(
    batchId: string,
    options: ImageBatchListOptions = {},
  ): Promise<ImageBatchResultPage> {
    return this.client.request<ImageBatchResultPage>({
      method: "GET",
      path: `/api/app/image-forensics/batches/${encodeURIComponent(batchId)}/results`,
      query: { skip: options.skip, take: options.take },
    });
  }

  /**
   * Cancel the images still waiting and refund their credits.
   *
   * Images already being analyzed are left to finish — they have consumed the work either way —
   * so the refund covers only what was actually cancelled.
   */
  async cancelBatch(batchId: string): Promise<ImageBatchCancelResult> {
    return this.client.request<ImageBatchCancelResult>({
      method: "POST",
      path: `/api/app/image-forensics/batches/${encodeURIComponent(batchId)}/cancel`,
      json: {},
    });
  }

  /**
   * Queue a batch and poll until it reaches a terminal state, then return it.
   *
   * **Rarely what you want.** A batch can legitimately take hours, so this holds a process open
   * for the entire window; the default poll timeout will usually expire first. Prefer
   * {@link Images.submitBatch}, persist the ID, and collect results later — that is the whole
   * point of the tier. This exists for tests and short-window setups.
   *
   * Polling stops on the server-computed `isTerminal` rather than a status allow-list, so it keeps
   * working if a status is ever added.
   */
  async submitBatchAndWait(
    files: FileInput[],
    options: ImageBatchSubmitOptions & PollOptions = {},
  ): Promise<ImageBatch> {
    const { intervalMs, timeoutMs, onPoll, ...submitOptions } = options;
    const { batchId } = await this.submitBatch(files, submitOptions);

    if (!batchId) {
      throw new Error("submitBatch did not return a batchId");
    }

    return pollUntil(
      () => this.getBatch(batchId),
      (batch) => batch.isTerminal === true,
      (batch) => batch.status ?? "unknown",
      { intervalMs, timeoutMs, onPoll },
    );
  }

  /**
   * Read the credit balance and the live image limits: cost per image, max file size and allowed
   * formats, plus whether the batch tier is available and what it costs. Read these rather than
   * hard-coding the defaults — they are account settings and can change without an SDK release.
   * Any valid API token may call this; no scope is required.
   */
  creditInfo(): Promise<ImageForensicsCreditInfo> {
    return this.client.request<ImageForensicsCreditInfo>({
      method: "GET",
      path: "/api/app/image-forensics/credit-info",
    });
  }
}
